code64.1
.TH code64 1 "\n[year]-\n[mo]-\n[dy]" "Linux"

.SH NAME
code64 - Base64 Encoder/Decoder Command Line Utility

.SH SYNOPSIS
.SY
.B code64
.RI [ OPTIONS ]
.RI [ filename ]
.YS

\# SECTION description
.SH DESCRIPTION
The \fBcode64\fR utility converts data to and from base64 encoding.

The source can be \fIstdin\fR or a filename, and the target can be
\fIstdout\fR or a filename.  The first non-option argument will be
taken as the input file, otherwise the input will be from \fIstdin\fR.
The output will be written to \fIstdout\fR unless the \fB-o\fR
option is used.

By default, the characters used for values 62 and 63 are '\fI+\fR'
and '\fI/\fR', and the padding character is '\fI=\fR'.  These characters
can be changed with options.

\# SECTION overview
.SH OVERVIEW
The \fBcode64\fR utility converts data to and from base64 encoding.  This
encoding is widely used to transmit files and data over the internet.
During encoding, every 3 8-bit bytes of input is converted to 4 6-bit
bytes of output.  The 6-bit variables represent values from 0 to 63,
and these values are represented by two runs of the roman alphabet,
first in upper case, then lower case, then the 10 numeric digits, 0-9.
This covers the first 62 possible values for a 6-bit value.  That
leaves 2 values unaccounted-for, 62 and 63.

In most cases, 62 and 63 are replaced with '\fI+\fR' and '\fI/\fR',
but there are exceptions.  URLs, for example, have specific
interpretations for the '\fI+\fR', which stands in for a space, and
'\fI/\fR' which normally indicates directory hierachy.

\# SECTION options
.SH OPTIONS
.TP
.BI -b " max_line_length"
.RS
Use this option to set the maximum line length of the encoded output.
The value will be coerced to the highest multiple of 4 that is equal
to or less than the requested line length.  A line break length of \fI0\fR
will omit all line breaks in the output.  The default value is \fI76\fR.
.RE
\#
.TP
.BI -c " string"
.RS
This option will set the special characters for values 62 and 63 and
for padding.  The string must be at least 2 characters long.  If the
string is 3 characters long, the third character will be used as a
padding character, if needed.  Otherwise, no padding will be used
when the final encoding group is less than 3 bytes long.  See
\fBSpecial Characters Warning\fR below.
.EX
\fBcode64 -c \fR '\fI+/=\fR'
.RS
Use '\fI+\fR' for 62, '\fI/\fR' for 63, and pad with '\fI=\fR'.
.RE
\fBcode64 -c \fR '\fI-_\fR'
.RS
Use '\fI-\fR' for 62, '\fI_\fR' for 63, and do not pad ending.
.RE
.EE
.RE
\#
.TP
.BI -d
.br
Decode the input file
\#
.TP
.BI -e
.br
Encode the input file
\#
.TP
.BI -h
.br
Present a short help page.
\#
.TP
.BI -i " input_file"
.br
Read from \fIinput_file\fR for encoding input instead of \fIstdin\fR.
\#
.TP
.BI -o " output_file"
.br
Write the encoded results to \fIoutput_file\fR instead of \fIstdout\fR.
\#
.TP
.BI -s " standard_name"
.br
This option serves as a shortcut to set the special characters and line
break length based on the intended standard invoked.

By default, \fBcode64\fR uses the \fIMIME\fR standard, using '\fI+\fR'
and '\fI/\fR' for values 62 and 63, '\fI=\fR' for padding, and a
maximum line length of 76 characters.  See \fISpecial Characters Warning\fR
and \fIOnline Reference\fR below.

The supported standards are:
.br
- pem
.br
- mime
.br
- rfc4648
.br
- base64url
.br
- radix64
.br
- utf-7
.br
- imap
.br
- y64
.br
- freenet
\# SECTION examples
.SH EXAMPLES
\#
.TP
.BI code64 " spreadsheet.xls"
Encode contents of file \fIspreadsheet.xls\fR and write to \fIstdout\fR.
\#
.TP
\fBcode64 -o\fI spreadsheet.b64 spreadsheet.ods\fR
Encode contents of \fIspreadsheet.ods\fR to \fIspreadsheet.b64\fR using
the MIME variation.
.TP
\fBecho -n\fI spreadsheet.ods\fR | \fBcode64\fR > \fIspreadsheet.b64\fR
Encode contents of \fIspreadsheet.ods\fR to \fIspreadsheet.b64\fR.

The same result as the previous example, but using pipes for input
and output.  In this case, note that \fBecho\fR uses the \fB-n\fR
option to omit the usual newline.  Otherwise, the newline character
would be included in the encoded result.
.TP
.BI "bcode64 -d " spreadsheet.b64 " -o " spreadsheet.ods
Decode contents of \fIspreadsheet.b64\fR to \fIspreadsheet.odx\fR.
.TP
.BI "bcode64 " spreadsheet.ods " -c " '#^=' " -b " 4
Encode contents of \fIspreadsheet.ods\fR, using '\fI#\fR' for 62,
'\fI^\fR' for 63, and '\fI=\fE' for padding, with 4 character lines
for debugging translation blocks.  See \fBSpecial Characters Warning\fR below.
\#
\# SECTION notes
.SH NOTES
.SS Special Characters Warning
The special characters in force during encoding must also be in force
during decoding.  Otherwise, the unrecognized special characters will
be ignored and the decoding will silently fail.

This is meant to be a feature, that the unrecognized characters are
ignored.  That's how multiple lines are treated as a single string.
However, the silent failure of mismatched special characters may pose
a greater problem.  It may be better to loudly fail in that case unless
a flag is set to ignore unrecognized characters.

.SS Online Reference
The reference I used to code and test this utility is at
.br
.UR https://en.wikipedia.org/wiki/Base64
.UE

The page also includes a table of encoding variants, with different
special characters, padding and line-length requirements.
.br
.UR https://en.wikipedia.org/wiki/Base64#Variants_summary_table
.UE
\#
\# SECTION see also
.SH SEE ALSO
.TP
.BR code64 "(3), " base64 (1)




code64.c
// -*- compile-command: "cc -Wall -Werror -O -I . -L. -o code64 code64.c  -Wl,-R -Wl,. -lcode64" -*-

#include <stdio.h>
#include <stdlib.h>   // for atoi();
#include <string.h>   // memset(), strerror()
#include <errno.h>    // make available the global errno variable

#include "code64.h"

typedef struct _Std_Type
{
   const char *name;
   const char *specials;
   unsigned int breaks;
} Std_Type;

Std_Type bstypes[9] = {
   { "pem",       "+/=",  64 },
   { "mime",      "+/=",  76 },
   { "rfc4648",   "+/=",   0 },
   { "base64url", "-_",    0 },
   { "radix64",   "+/=",  76 },
   { "utf-7",     "+/",    0 },
   { "imap",      "+,",    0 },
   { "y64",       "._-",   0 },
   { "freenet",   "~-=",   0 }
};

unsigned int number_of_bstypes = sizeof(bstypes) / sizeof(Std_Type);

void show_standards(void)
{
   const Std_Type *ptr = bstypes;
   for (unsigned int i=0; i < number_of_bstypes; ++i)
   {
      printf("   %s\n", ptr->name);
      ++ptr;
   }
}

void show_usage(void)
{
   printf("-b line length  Number of characters per line in encoded output.\n");
   printf("-c set custom special characters.\n");
   printf("-d to decode file.\n");
   printf("-e to encode file without breaks.\n");
   printf("-h *help* to show usage (this display).\n");
   printf("-i filename Read filename instead of reading stdin for input.\n");
   printf("-o filename Write to filename instead to stdout.\n");
   printf("-s standard to use for special characters, padding, and line length.\n");
   printf("   The following standards are recognized:\n");
   show_standards();
}

void close_FILEs(FILE *file1, FILE *file2)
{
   if (file1 != NULL)
      fclose(file1);
   if (file2 != NULL)
      fclose(file2);
}

int set_special_chars_from_string(const char *str)
{
   if (strlen(str) > 1)
   {
      c64_set_special_chars(str);
      return 1;
   }
   else
      return 0;
}

const Std_Type* get_standard(const char *sname)
{
   const Std_Type *ctype = bstypes;

   for (unsigned int i=0; i < number_of_bstypes; ++i)
   {
      if (0 == strcmp(ctype->name, sname))
         return ctype;
      ++ctype;
   }
   return 0;
}

/**
 * @brief Return a valid breaks value (divisble by 4).  Returns 0 if less than 3.
 */
int align_breaks_value(const char *val)
{
   int value = atoi(val);
   if (value > 3)
      return value / 4 * 4;
   else
      return 0;
}

int main(int argc, const char **argv)
{
   enum ops { None, Encode, Decode };

   // FILE stream pointers to be used for input and output.
   // Although they may point to different streams, they will
   // NOT be closed using these pointers.
   FILE *fin_using = stdin;
   FILE *fout_using = stdout;

   // These variables will point to command line arguments if
   // the -i and/or -o options are used.  Non-null values below
   // will trigger a call to fopen() when the option processing
   // is done.
   const char *in_filename = NULL;
   const char *out_filename = NULL;

   // These variable will be NULL unless -i or -o has set
   // the in_filename or out_filename values, in which case
   // these variables will hold the pointers to the opened
   // streams.  The fin_using and fout_using NULL values will
   // be replaced with these values when and it these pointers
   // are set.  These stream pointers WILL be closed at the
   // end if they are not NULL.
   FILE *fin = NULL;
   FILE *fout = NULL;

   const Std_Type *selected_stype = NULL;

   enum ops operation = Encode;
   int breaks = 76;

   if (argc == 1)
   {
      show_usage();
   }
   else
   {
      const char **ptr = argv;
      int count = 0;
      while (count < argc)
      {
         if (count)
         {
            // If not an option, assume the argument is the infile name:
            if (**ptr != '-')
               in_filename = *ptr;
            else
            {
               switch((*ptr)[1])
               {
                  case 'b':
                     ++ptr;
                     ++count;
                     breaks = align_breaks_value(*ptr);
                     if (breaks)
                        fprintf(stderr, "Breaking lines at the %d character.\n", breaks);
                     else
                        fprintf(stderr, "Omitting line breaks in output.\n");
                     break;
                  case 'c':
                     ++ptr;
                     ++count;
                     if (!set_special_chars_from_string(*ptr))
                        fprintf(stderr, "Special characters string too short.\n");
                     break;
                  case 'd':
                     operation = Decode;
                     break;
                  case 'e':
                     operation = Encode;
                     break;
                  case 'h':
                     show_usage();
                     return 0;
                  case 'i':
                     ++ptr;
                     ++count;
                     in_filename = *ptr;
                     break;
                  case 'o':
                     ++ptr;
                     ++count;
                     out_filename = *ptr;
                     break;
                  case 's':
                     ++ptr;
                     ++count;
                     if ((selected_stype = get_standard(*ptr)))
                     {
                        set_special_chars_from_string(selected_stype->specials);
                        breaks = selected_stype->breaks;
                     }
                     else
                        fprintf(stderr, "Unrecognized standard name '%s'.\n", *ptr);
                     break;
                  default:
                     show_usage();
                     return 1;
               }
            }
         }

         ++count;
         ++ptr;
      }

      if (in_filename)
      {
         fin = fopen(in_filename, "r");
         if (fin)
            fin_using = fin;
         else
         {
            fprintf(stderr, "Failed to open input file \"%s\" (%s).\n", in_filename, strerror(errno));
            close_FILEs(fin, fout);
            return 1;
         }
      }

      if (out_filename)
      {
         fout = fopen(out_filename, "w");
         if (fout)
            fout_using = fout;
         else
         {
            fprintf(stderr, "Failed to open out file \"%s\" (%s).\n", out_filename, strerror(errno));
            close_FILEs(fin, fout);
            return 1;
         }
      }

      if (operation == Encode)
         c64_encode_stream_to_stream(fin_using, fout_using, breaks);
      else if (operation == Decode)
         c64_decode_stream_to_stream(fin_using, fout_using);

      close_FILEs(fin, fout);
   }

   return 0;
}
